File: linux/seek_hole.py
#!/usr/bin/env python3
# Iterate over the holes in a file
# https://claude.ai/chat/07491d10-2a43-4469-a82f-51446c38bab4

import os, errno, sys
fd = os.open(
    sys.argv[1],
    os.O_RDONLY)
end = os.fstat(fd).st_size
pos = 0
while pos < end:
    try:
        data = os.lseek(fd, pos, os.SEEK_DATA)
    except OSError:
        break  # only holes remain
    hole = os.lseek(fd, data, os.SEEK_HOLE)
    print(f"data: {data}..{hole} n={hole-data}")
    pos = hole

File: linux/seek_hole.c
#include "macros.h"
int main(int argc, char *argv[])
{
    if (argc != 2) {
        ERROR("usage: %s <file>\n", argv[0]);
        return 1;
    }
    int fd;
    ASSERT_ERRNO(fd = open(argv[1], O_RDONLY));
    struct stat st;
    ASSERT_ERRNO(fstat(fd, &st));
    off_t end = st.st_size;

    off_t pos = 0;
    while (pos < end) {
        off_t data = lseek(fd, pos, SEEK_DATA);
        if (data < 0) {
            if (errno != ENXIO) {
                ERROR("lseek SEEK_DATA");
            }
            break;  /* ENXIO: only holes remain */
        }
        off_t hole = lseek(fd, data, SEEK_HOLE);
        if (hole < 0) {
            ERROR("lseek SEEK_HOLE");
            break;
        }
        long long hole_size = hole-data;
        LOG("data: %lld..%lld n=%lld\n",
            (long long)data, (long long)hole, hole_size);
        pos = hole;
    }

    close(fd);
    return 0;
}
